### ISSUE TITLE
Whack-A-Mole VR Dashboard: Data capture experiment

### ISSUE THREAD

**Issue by bastianilso** [med-material/r-shiny-js-data-capture#13]
Once the data capture system is ready in its first version (after solving issue #5), we should make a data collection test with it.
We should test the data capture system to verify that we can collect data from a real dashboard.

Steps:
1. Embed the JS data capture system with the Whack-A-Mole Dashboard at https://github.com/med-material/Whack_A_Mole_RShiny
2. Run experimental procedure described below with 1-2 persons.
3. Import collected data into R and analyze it, based on the analysis goals described below.

Experimental Procedure:
- Place a laptop on a table, with the Whack-A-Mole Dashboard + a screen recorder (I recommend to use OBS Screen recorder)
- Get a person to sit in front of the laptop and ask the person to solve 3 tasks. For each task, start the screen recording and the data capture system, and stop it again.
- The 3 tasks to solve are: 1) Upload data to the dashboard, 2) Find out whether Game Performance is worse on the left side than the right side and 3) Find out at what point in the game, the person missed moles.

Analysis Goals:
We will use the data to:
- Identify which task took the longest time to solve.
- Count how many clicks it took to solve each task.
- Measure how much distance, the person's mouse had to move in each task.
- Measure how much the person had to scroll.
- Visualize the trajectory of the person's mouse movements.

**Comment by bastianilso** [comment]
Data to upload (upload in the order: Meta, Event, Sample)
https://github.com/med-material/d3-rshiny-vis/files/9958725/log2021-07-01.12-50-01.9658Sample.zip

**Comment by bilal-62210** [comment]
@bastianilso i've send on your email address the test result because the folder is too big to be send here

**Comment by bastianilso** [comment]
OK @bilal-62210, now focus on the analysis goals and help provide answer to the following questions. I suggest you try to do it in a new project in R.


>     Identify which task took the longest time to solve.
>     Count how many clicks it took to solve each task.
>     Measure how much distance, the person's mouse had to move in each task.
>     Measure how much the person had to scroll.
>     Visualize the trajectory of the person's mouse movements.

**Comment by bastianilso** [comment]
@bilal-62210 i noticed in aldryck's "step1" data, it says the duration is "87065" (seconds?)
![image](https://user-images.githubusercontent.com/3967945/205267289-d6b4d534-c2f1-4315-af58-e5cf8d280be2.png)

that's a either a very long time, or maybe a count of milliseconds (?) 
I think it would be better to write "87.065" seconds then. (this corresponds to the `ss.fff` format we know  from timestamps `hh:mm:ss.fff`

**Comment by bastianilso** [comment]
@bilal-62210 another suggestion, in your analysis project, after you import the CSV files you created with LoggingLoader and save the data as a single RDA file.

You can see how saving/loading RDA files in your R project works here:
https://github.com/bastianilso/bci-pam-analysis/blob/main/pam_study_preprocess.R#L16

in general, feel free to take inspiration from the general structure of analysis from here:
https://github.com/bastianilso/bci-pam-analysis

we use a notion of having a "preprocess" file and an "analysis" file - in your case, an "analyis" file is probably enough for now.
preprocessing is mainly used in case the data needs to be cleaned (formatting timestamps, formatting numbers..)

**Comment by bilal-62210** [comment]
> @bilal-62210j'ai remarqué dans les données "step1" d'aldryck, il est indiqué que la durée est de "87065" (secondes ?) dans le méta-fichier ![image](https://user-images.githubusercontent.com/3967945/205267289-d6b4d534-c2f1-4315-af58-e5cf8d280be2.png)
> 
> c'est soit très longtemps, soit peut-être un nombre de millisecondes (?) Je pense qu'il serait préférable d'écrire "87,065" secondes alors. (cela correspond au `ss.fff`format que nous connaissons des horodatages`hh:mm:ss.fff`

Normaly there is a point in the value like this: 
![image](https://user-images.githubusercontent.com/73820951/205281469-08ff6a4a-aec1-4450-a184-3d34552855e6.png)

**Comment by bastianilso** [comment]
@bilal-62210 yes, this solved it, i have the point too.

**Comment by bilal-62210** [comment]
@bastianilso this is what i have for the duration of a task : 

![image](https://user-images.githubusercontent.com/73820951/208081308-34db84c8-1432-4942-aeed-b91791621cbe.png)

So for this analysis goal : Identify which task took the longest time to solve, we just have to compare the duration of each task to know wich tasj took the longest time to solve.

**Comment by bastianilso** [comment]
@bilal-62210 Thanks. Here is some feedback on your process:

- You are writing code directly in the R terminal, I think it's better to use an empty R project and write code in there, so you can easily update it.
- You are loading data from an absolute path, this should not be necessary. You can ask andy/aldryck/pierre about help for that if need be. Your r project should contain a subfolder called "data" and thus you can load the data by specifying `LoadFromDirectory("data/")`.  Make sure to pass the parameter `Sample="ContinuousMeasurement"` since your continuous measurements are called that.
- Your experiment collected data from 2 people (Aldryck/Xavier) in 3 tasks. Therefore I would expect **2x3=6** indications of duration (one duration for each task for each participant). Right now the Duration column only contains one number (5.167) and its not clear which person or which task the number belongs to. What data is  inside the data folder?
-  Judging from the terminal output, the Duration column currently holds data in "chr" format (character format). You may want to format that as "num" format (numerical) (look this up online).
- Whether performing a "Distinct()" operation on D is the right choice depends what you are trying to do. Distinct removes all non-unique numbers. But what you actually might want to do is to use `group_by()` and group by the columns which indicate which participant and test this data is from. in addition, you may want to use a `summarize()` function to summarize the duration by each. There are youtube videos demonstrating how to do this with dplyr online.
- The comparison of duration should ideally not be manual. If you were to code a dashboard like whack-a-mole, you would want to programmatic way to identify which duration was the shortest.

**Comment by bilal-62210** [comment]
hi @bastianilso, i'm working on the analaysis goals and i've did a wrong manipulation, i've deleted the folder with xavier's and aldryck's test. Do you still have this folder ? if yes please send it to me. If you don't have it i will do some new tests. Thanks

**Comment by bastianilso** [comment]
Hi @bilal-62210,

I have it on my work laptop. Better do some new tests. Use 2 other participants so we can combine the data with the data from xavier and aldryck when i come back. -B

**Comment by bilal-62210** [comment]
@bastianilso ok I will do it with Pierre and Lucas.

**Comment by bilal-62210** [comment]
@bastianilso i learned a lot about dplyr the last 2 days, this is my result for the duration of a task : 
![image](https://user-images.githubusercontent.com/73820951/209161661-b4500949-913e-40e2-9207-8d5262b3ecfd.png)
![image](https://user-images.githubusercontent.com/73820951/209161745-3285863a-847f-4868-9412-5e532c53f750.png)
Give me your feedback about that

**Comment by bastianilso** [comment]
nice @bilal-62210, yes these results allow us to compare the duration and solve the first analysis goal. Here are some notes from me about your solution:

1) your dataset 'D' contains Duration.x and Duration.y because the Event CSV and Meta CSV each contain a column called Duration. To fix this, I suggest we rename the Meta CSV duration column to e.g ."SessionDuration" and the Event duration to e.g. "DurationSinceStart". This also makes it more clear what the columns contain than the word Duration itself. I will create a separate issue about that.

2) As a preliminary step to your analysis you can consider renaming the columns "i2" and "i3" to e.g. "Participant" and "Task". you can use dplyr's `rename()` function to do this.

3) Rather than storing the results as intermediate variables you can consider concatenating line 6-10 like this (assuming the variables are renamed as per suggestions above):

```
DurationSummary = D %>% group_by(Participant, Task) %>%
  mutate(SessionDuration = as.numeric(SessionDuration)) %>%
  summarize(MeanDuration = mean(SessionDuration)) %>%
  arrange(desc(MeanDuration))
```
If you are in doubt how to see intermediate results, you can apply `%>% view()` at any stage of the dplyr pipeline and R studio will show you the dataset in its state at that point. For example `D %>% group_by(Participant, Task) %>%
  mutate(SessionDuration = as.numeric(SessionDuration)) %>% view()` will view the dataset after R has performed the group_by and mutate operation on D.

**Comment by bilal-62210** [comment]
@bastianilso Ok i will work on it and make the changes.
I've answered to your first point in the issue #16.

**Comment by bilal-62210** [comment]
@bastianilso Hi, i've worked on this analysis goal : "Count how many clicks it took to solve each task" 
And this is my result: 
![image](https://user-images.githubusercontent.com/73820951/210567480-38d92d23-09b5-4b97-90cb-38bf2c5b4be5.png)
What do you think about that? I think it can be smart to count all events and not only de number of click

**Comment by bastianilso** [comment]
lead/lag values might help, can be used inside a mutate function:
https://dplyr.tidyverse.org/reference/lead-lag.html
